crates/services/src/seqlock.rs

```diff
@@ -0,0 +1,165 @@
+//! A simple implementation of a sequential lock.
+//! More details: <https://docs.kernel.org/locking/seqlock.html>
+
+use std::{
+    cell::UnsafeCell,
+    panic::UnwindSafe,
+    sync::atomic::{
+        AtomicU64,
+        Ordering,
+    },
+};
+
+/// A simple implementation of a sequential lock.
+/// some usage of unsafe, T must be Copy
+#[derive(Debug)]
+pub struct SeqLock<T: Copy> {
+    sequence: AtomicU64,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send + Copy> Sync for SeqLock<T> {}
+
+/// The writer handle for the `SeqLock`.
+/// Only one writer exists for a `SeqLock`.
+/// There is no Clone bound since we want to enforce only one writer.
+#[derive(Debug)]
+pub struct SeqLockWriter<T: Copy> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T: Copy> SeqLockWriter<T> {
+    /// Modifies the data within the lock.
+    pub fn write<F>(&self, f: F)
+    where
+        F: FnOnce(&mut T) + UnwindSafe,
+    {
+        let lock = &self.lock;
+
+        // Indicate that a write operation is starting.
+        lock.sequence.fetch_add(1, Ordering::Release);
+
+        // attempt to perform the write, and catch any panics
+        // safety: panics are caught and resumed
+        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| unsafe {
+            let data = &mut *lock.data.get();
+            f(data);
+        }));
+
+        // Indicate that the write operation has finished.
+        lock.sequence.fetch_add(1, Ordering::Release);
+
+        // resume unwinding if there was an error
+        if let Err(e) = result {
+            std::panic::resume_unwind(e);
+        }
+    }
+}
+
+/// The reader handle for the `SeqLock`.
+/// Multiple readers can be created for a `SeqLock`.
+#[derive(Clone, Debug)]
+pub struct SeqLockReader<T: Copy> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T: Copy> SeqLockReader<T> {
+    /// Reads the data within the lock.
+    pub fn read(&self) -> T {
+        let lock = &self.lock;
+
+        loop {
+            // check starting guard
+            let start = lock.sequence.load(Ordering::Acquire);
+
+            // if odd, write in progress
+            if start % 2 != 0 {
+                continue;
```

Comment:
We should probably call [`std::thread::yield_now`](https://doc.rust-lang.org/std/thread/fn.yield_now.html) before `continue`. See https://github.com/Amanieu/seqlock/blob/b72653fd9c38141da7e588a97a1b654239f5d4df/src/lib.rs#L126

Reply:
addressed in 51d079ca61